### ISSUE TITLE
qt5reactor crashes when used with twisted.internet.serialport.SerialPort

### ISSUE THREAD

**Issue by GMIG** [twisted/qt5reactor#37]
I am making a serial port monitor with PyQT5 and Twisted on Windows and have encounted presumably a QT5 internal error. This is the code:

    import sys
    from PyQt5.QtWidgets import QApplication, QWidget
    app = QApplication(sys.argv)
    import qt5reactor
    qt5reactor.install()

    from twisted.internet import reactor
    from twisted.protocols.basic import LineReceiver
    from twisted.internet.serialport import SerialPort
    
    
    class Myform(QWidget):
        def __init__(self):
            super().__init__()
            self.setGeometry(10, 10, 100, 100)
            self.show()
    
    
    class Prot(LineReceiver):
        def rawDataReceived(self, data):
            pass
    
        def lineReceived(self, line):
            pass
    
    
    s = SerialPort(Prot(), "com3", reactor, baudrate=57600)
    ex = Myform()
    reactor.run()

This gives the following error

    Process finished with exit code -1073740791 (0xC0000409)

If the `SerialPort ` initialization string is commented, everything works perfectly (the window appears). Otherwize if `qt5reactor.install()` is commented the twisted code works normally but QT app (obviously) isn't started.

**Comment by altendky** [comment]
I really should have thought of this as soon as you submitted...  Try enabling the [`faulthandler` module](https://docs.python.org/3/library/faulthandler.html) and see if we get a useful C/C++/Python hybrid stack trace.

**Comment by mss88** [comment]
in qt5reactor/core.py -> class QtEventReactor(QtReactor) ->function doEvents(self)
1. `handles = self._events.keys()` returns <dict_keys> object, not list
2. `val = MsgWaitForMultipleObjects(handles, 0, 0, QS_ALLINPUT | QS_ALLEVENTS)` function requires list of integers (actually list of <pyHANDLE> objects)
3. i've changed  `handles = self._events.keys()` to `handles = list(self._events.keys())`
4. serialport recive works with GUI (windows 7 64 bit, python 3.7) but only after another task.LoopingCall calls cb-function

**Comment by altendky** [comment]
There's that change in PR form.  Can you confirm that's what you meant?  I'm not sure what 4) was about though.  Only after another `task.LoopingCall` calls what?  What is the first `LoopingCall`?

**Comment by mss88** [comment]
`def success(args, hostname, window):
    (errorStatus, errorIndex, varBinds) = args

    str_status = ""

    if errorStatus:
        str_status  = '%s: %s at %s' % (hostname,
                                errorStatus.prettyPrint(),
                                errorIndex and varBinds[int(errorIndex) - 1][0] or '?')
    else:
        for varBind in varBinds:
            str_status = ' = '.join([x.prettyPrint() for x in varBind])

    window.snmpTextField.setText(str_status)

def failure(errorIndication, hostname):
    print('%s failure: %s' % (hostname, errorIndication))


class SerialProtocol(Protocol):
    window: 'MainWindow'

    def connectionMade(self):
        print('Serial Connection made!')

    def connectionLost(self, reason):
        print('Connection lost!', reason)

    def dataReceived(self, data):
        print("Got Serial data:")
        window.serialTextField.setText(repr(data))

    def sendData(self, tx_data):
        print('write to serial port:', repr(tx_data))
        try:
            self.transport.write(tx_data.encode())
        except:
            print('cant write to serial port')
            pass


if __name__ == '__main__':
    app = QApplication(sys.argv)

    qt5reactor.install()

    window = MainWindow()
    window.show()

    from twisted.internet import reactor, task, threads
    from pysnmp.hlapi.twisted import *

    def getSysUpTime(reactor, hostname):
        d = getCmd(SnmpEngine(),
                   CommunityData('public', mpModel=0),
                   UdpTransportTarget((hostname, 161)),
                   ContextData(),
                   ObjectType(ObjectIdentity('SNMPv2-MIB', 'sysUpTime', 0)))

        d.addCallback(success, hostname, window).addErrback(failure, hostname)
        return d

    serial_protocol = SerialProtocol()
    sp = SerialPort(serial_protocol, 'COM43', reactor, baudrate='9600')

    window.reactor = reactor
    window.serial_protocol = serial_protocol

    l = task.LoopingCall(getSysUpTime, reactor, "192.168.0.145")
    l.start(2.0)

    reactor.run()
`
here is part of my code. I want to periodically interrogate snmp device and send result data to arduino by serial port. serial_protocol recieves data from arduino after snmp succes of failure callbacks.

**Comment by altendky** [comment]
Does #49 fix it completely or do you still feel like there's a problem here?  I'm still not clear what you mean by "but only after another task.LoopingCall calls cb-function".

You can install #49 as like `yourenv/scripts/pip install git+https://github.com/sunu/qt5reactor@list_for_pywin32#egg=qt5reactor` to test it out.

**Comment by altendky** [comment]
@mss88 and @GMIG, might you have a chance to check this?  With a little luck we'll be making a release 'soon' and it would be nice to include a fix for your problem.

updated installation command: `yourenv/scripts/pip install git+https://github.com/twisted/qt5reactor@list_for_pywin32#egg=qt5reactor`

**Comment by altendky** [comment]
Closing as I merged and will release #49.  If the same issue remains feel free to reopen (or ask me to, not sure about permissions here).
